### PR TITLE
Per-geometry OHEM vol_pressure loss weighting for OOD hard cases

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Four test geometries (run_133, run_226, run_203, run_158) account for approximately 92% of the squared volume_pressure error, creating a ~3.2× val/test gap (val≈3.6%, test≈11.3%). The existing training loss treats all geometries equally, so the model never sees a strong signal to learn the rare geometric patterns that these OOD cases expose.

Per-sample loss weighting at training time can bias the optimizer toward better representations of OOD-like geometry. The hypothesis: if the model already misrepresents certain geometric families, we can detect this using the validation error residuals and up-weight training samples with similar geometric signatures, improving generalization to the OOD test cases.

**Specifically:** compute per-geometry validation volume_pressure L2 error at the end of each epoch. Identify the top-K highest-error training geometries (by geometric similarity to the worst-performing val geometries, approximated via surface area, drag coefficient bucket, or random ID assignment). Up-weight their training loss by a factor of α in the following epoch. This is an online hard-example mining approach (OHEM), adapted for geometry-level batching.

## Instructions

Add per-geometry volume-pressure OHEM loss reweighting to `train.py`:

1. **Add a `GeometryOHEM` class** that tracks per-case volume_pressure L2 error across the validation set. After each validation pass:
   - Store `{case_id: vol_pressure_l2}` for every geometry evaluated
   - Rank geometries by vol_pressure error (descending)
   - Build a weight dict: top-20% hardest geometries get weight α, rest get weight 1.0
   - α is controlled by `--ohem-alpha` (default: 3.0)

2. **Pass per-sample weights to the training loss**: in the training loop, look up `ohem_weights[case_id]` for each sample in the batch, multiply the volume_pressure component of the per-sample loss by this weight before summing/averaging.

3. **Add new CLI flags**:
   - `--use-vol-ohem` (bool, default False) — enables this feature
   - `--ohem-alpha FLOAT` (default 3.0) — hard-sample up-weighting factor
   - `--ohem-hard-frac FLOAT` (default 0.2) — fraction of cases considered "hard"

4. **Warmup**: For the first epoch (before any val pass), use uniform weights = 1.0. After the first val pass, apply OHEM weights for subsequent epochs.

5. **Log OHEM weights** to W&B each epoch: `wandb.log({"ohem/max_weight": ..., "ohem/mean_weight": ..., "ohem/hard_case_count": ...})`

Run with full SOTA config + OHEM:

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent tanjiro --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 13 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --use-vol-ohem --ohem-alpha 3.0 --ohem-hard-frac 0.2 \
  --wandb-group tanjiro-ohem-vol-pressure \
  --wandb-name tanjiro/ohem-alpha3-frac0.2-run1
```

Also try a second run with α=5.0 to test sensitivity:

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent tanjiro --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 13 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --use-vol-ohem --ohem-alpha 5.0 --ohem-hard-frac 0.2 \
  --wandb-group tanjiro-ohem-vol-pressure \
  --wandb-name tanjiro/ohem-alpha5-frac0.2-run2
```

Run both sequentially if GPUs allow, or choose the most promising α first. Report val_abupt and **test vol_pressure** for both runs.

## Baseline (current SOTA to beat)

| Metric | Val | Test |
|---|---:|---:|
| abupt | **6.4407%** | 7.6992% |
| surface_pressure | 4.1836% | 3.8451% |
| volume_pressure | 3.8557% | **11.6704%** |
| wall_shear | 7.3448% | 7.0429% |

**Single-model training gate:** val_abupt < **6.4407%**  
**W&B run (SOTA):** `ghh0s4ne` (PR #823, group `nezuko-surf-vol-xattn`)  
**W&B project:** `wandb-applied-ai-team/senpai-v1-drivaerml-ddp8`

**Primary target:** volume_pressure test error — currently 11.67%, target 6.08% (AB-UPT)  
**OOD note:** 4 test geometries (run_133, run_226, run_203, run_158) drive ~92% of squared vol error.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #904 because ADVISOR: Closing — this experiment duplicates PR #888 (thorfinn OOD upweighting ×3, CLOSED NEGATIVE). That experiment reached the definitive conclusion: **the OOD gap is not data-distribution-limited; it cannot be bridged by train-set reweighting**. The 4 OOD test geometries (run_133, run_226, run_203, run_158) are test-side only. Even with nearest-train-neighbor upweighting (×3) the forced-timeout val_abupt was 7.42% (+0.98pp above SOTA 6.44%) and test_abupt was 8.69% (+2.25pp). OHEM at any α would trigger the same fundamental failure: the train distribution does not contain those geometries. tanjiro is already assigned to PR #883 (xattn pos-encoding), which remains the active WIP. Closing this PR to avoid confusion..
